Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -88,32 +88,122 @@ extension Manifest {
 }
 
 public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
-
+    
+    // FIXME: Remove this property and the initializer?
```

Comment:
I'd have to check, but I thought this was used in other places in the code as well, in order to know the latest version (for things like diagnostic messages, if nothing else).  Conceptually it makes sense for the SwiftPM code base to know its current version.

Reply:
Actually I see what you mean now.  The usual one is `ToolsVersion.currentToolsVersion`.  Yes, in that case, this one can probably be removed.

Reply:
I think the only thing that prevents `ToolsVersionLoader.currentToolsVersion` from being removed is [`ToolsVersionLoaderTests.testVersionSpecificManifestFallbacks()`](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/d433487e6461821c4e7d2c8cbbac4cdc2f0c1252/Tests/PackageLoadingTests/ToolsVersionLoaderTests.swift#L642). It uses the property to check that SwiftPM chooses a version-specific manifest correctly.

[`ToolsVersionLoader.load(at:fileSystem:)`](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/d433487e6461821c4e7d2c8cbbac4cdc2f0c1252/Sources/PackageLoading/ToolsVersionLoader.swift#L204) and [`Manifest.path(atPackagePath:currentToolsVersion:fileSystem)`](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/d433487e6461821c4e7d2c8cbbac4cdc2f0c1252/Sources/PackageLoading/ToolsVersionLoader.swift#L34) use this property too. But this is only because `ToolsVersionLoaderTests.testVersionSpecificManifestFallbacks()` needs it.

What should I do to move the logic completely into the test file?

Reply:
Having looked into this a little, it seems a bit more complicated than I thought.  Unlike `ToolsVersion.currentVersion`, it seems as if this one really is intended to allow the simulation of other tools versions than the one that is actually being built, just for testing purposes.  So I don't think that removing it should be part of this PR, and having looked it, it is possible that it shouldn't actually be removed if it provides a necessary testability hook.  But it would have helped if there had been a comment there pointing out its purpose, so if you want to address anything for this PR, adding such a comment (that it's a testability parameter) might be warranted.  I am not that familiar with this code but it's possible that this could be refactored into protocols that provide that testability without requiring that additional parameter.  But I don't think it should hold up this PR.

Reply:
I left some additional comment on this, explaining what the property is for, and linked to this discussion. This should help future readers understand it better.

Reply:
Thank you, that's very helpful!